neo/app/controllers/Cloudberry.scala

```diff
@@ -0,0 +1,229 @@
+import edu.uci.ics.cloudberry.zion.model.impl._
+import play.Logger
+import play.api.data.Form
+import play.api.data.Forms._
+import play.api.libs.concurrent.Execution.Implicits.defaultContext
+import play.api.libs.json.{JsSuccess, JsValue, _}
+import play.api.libs.streams.ActorFlow
+import play.api.libs.ws.WSClient
+import play.api.mvc.{Action, Controller, WebSocket}
+import play.api.{Configuration, Environment}
+import play.api.i18n.{I18nSupport, MessagesApi}
+
+import scala.concurrent.duration._
+import scala.concurrent.{Await, Future}
+
+@Singleton
+class Cloudberry @Inject()(val wsClient: WSClient,
+                           val configuration: Configuration,
+                           val messagesApi: MessagesApi,
+                           val environment: Environment)
+                          (implicit val system: ActorSystem,
+                           implicit val materializer: Materializer
+                          ) extends Controller with I18nSupport {
+  val config = new Config(configuration)
+  val (asterixConn, qlGenerator) =
+    config.AsterixLang match {
+      case "aql" => (new AsterixAQLConn(config.AsterixURL, wsClient), AQLGenerator)
+      case "sqlpp" => (new AsterixSQLPPConn(config.AsterixURL, wsClient), SQLPPGenerator)
+      case _ => throw new IllegalArgumentException(s"unknown asterixdb.lang option:${config.AsterixLang}")
+    }
+
+  Await.result(Migration_20160814.migration.up(asterixConn), 10.seconds)
+
+  val manager = system.actorOf(DataStoreManager.props(Migration_20160814.berryMeta, asterixConn, qlGenerator, config))
+
+  Logger.info("Cloudberry is initializing")
+
+  val listener = system.actorOf(Props(classOf[Listener], this))
+  system.eventStream.subscribe(listener, classOf[DeadLetter])
+
+  def index = Action.async {
+    implicit val askTimeOut: Timeout = config.UserTimeOut
+    (manager ? DataStoreManager.ListAllDataset).map { case dataset: Seq[DataSetInfo] =>
+      Ok(views.html.index(dataset.filter(_.createQueryOpt.isEmpty).map(DataSetInfo.write), Cloudberry.createRegisterForm))
+    }
+  }
+
+  def webRegister = Action.async { implicit request =>
+    import Cloudberry.createRegisterForm
+    val formValidationResult = createRegisterForm.bindFromRequest
+    formValidationResult.fold({ formWithErrors =>
+      // This is the bad case, where the form had validation errors.
+      // Let's show the user the form again, with the errors highlighted.
+      // Note how we pass the form with errors to the template.
+      implicit val askTimeOut: Timeout = config.UserTimeOut
+      (manager ? DataStoreManager.ListAllDataset).map { case dataset: Seq[DataSetInfo] =>
+        BadRequest(views.html.index(dataset.map(DataSetInfo.write), formWithErrors))
+      }
+    }, { registerForm =>
+      // This is the good case, where the form was successfully parsed as a Widget.
+      val json = Json.parse(registerForm.registerJSONString)
+      json.validate[Register] match {
```

Comment:
From line 76 - line 96 are the same as in `register`, maybe we can write it in a function and call?